File: DataStructure/LinkedList/singly_linked_list.py
class Node:
    def __init__(self, item):
        self.val = item
        self.next = None


class LinkedList:
    def __init__(self, item):
        self.head = Node(item)

    def add(self, item):
        cur = self.head
        while cur.next is not None:
            cur = cur.next
        cur.next = Node(item)

    def remove(self, item):
        if self.head.val == item:
            self.head = self.head.next
        else:
            cur = self.head
            while cur.next is not None:
                if cur.next.val == item:
                    next = cur.next.next
                    cur.next = next
                else:
                    cur = cur.next

    def reverse(self):
        cur = self.head
        prev = None
        while cur:
            next = cur.next
            cur.next = prev
            prev = cur
            cur = next
        self.head = prev

    def print_list(self):
        cur = self.head
        while cur:
            print(cur.val)
            cur = cur.next


ll = LinkedList(1)
ll.add(2)
ll.add(3)
ll.add(4)
ll.add(5)
ll.add(6)
ll.remove(6)
ll.reverse()
ll.print_list()

File: DataStructure/LinkedList/singlyLinkedList.dart
class Node {
  int? value;
  Node? next;

  Node(this.value);
}

class LinkedList {
  late Node head;
  late Node tail;
  int length = 0;

  LinkedList() {
    head = Node(null);
    tail = head;
  }

  bool isEmtpy() {
    if (head.value == null) {
      return true;
    }

    return false;
  }

  void add(int value) {
    // if (head.value == null && tail.value == null) {
    if (length == 0) {
      head = Node(value);
      tail = head;
    } else {
      Node currentTail = tail;
      tail = Node(value);
      currentTail.next = tail;
    }

    length += 1;
  }

  bool remove(int index) {
    if (index >= length) {
      print('wrong index');
      return false;
    }

    if (index == 0) {
      if (length == 1) {
        head = Node(null);
        tail = head;
      } else {
        head = head.next!;
      }
    } else {
      int idx = 0;
      Node currentNode = head;

      while (idx < index - 1) {
        currentNode = currentNode.next!;
        idx += 1;
      }

      currentNode.next = currentNode.next!.next;

      if (currentNode.next == null) {
        tail = currentNode;
      }
    }

    length -= 1;
    return true;
  }

  void printLinkedList() {
    Node? currentNode = head;
    int idx = 0;

    while (idx < length) {
      print(currentNode!.value);
      currentNode = currentNode.next;
      idx += 1;
    }
  }

  void reverse() {
    if (length > 1) {
      Node? prevNode = null;
      Node? currentNode = head;
      tail = currentNode;

      while (currentNode != null) {
        Node? nextNode = currentNode.next;
        currentNode.next = prevNode;
        prevNode = currentNode;
        currentNode = nextNode;
      }

      head = prevNode!;
    }
  }
}

void main() {
  LinkedList linkedList = LinkedList();
  linkedList.add(1);
  linkedList.add(2);
  linkedList.add(3);
  linkedList.add(4);
  linkedList.add(5);
  
  linkedList.reverse();
  linkedList.printLinkedList();
}